freqtrade/data/dataprovider.py

```diff
@@ -63,9 +79,99 @@ def _set_cached_df(
         :param dataframe: analyzed dataframe
         :param candle_type: Any of the enum CandleType (must match trading mode!)
         """
-        self.__cached_pairs[(pair, timeframe, candle_type)] = (
+        pair_key = (pair, timeframe, candle_type)
+        self.__cached_pairs[pair_key] = (
             dataframe, datetime.now(timezone.utc))
 
+    # For multiple producers we will want to merge the pairlists instead of overwriting
+    def set_producer_pairs(self, pairlist: List[str], producer_name: str = "default"):
+        """
+        Set the pairs received to later be used.
+        This only supports 1 Producer right now.
+
+        :param pairlist: List of pairs
+        """
+        self.__producer_pairs[producer_name] = pairlist.copy()
+
+    def get_producer_pairs(self, producer_name: str = "default") -> List[str]:
+        """
+        Get the pairs cached from the producer
+
+        :returns: List of pairs
+        """
+        return self.__producer_pairs.get(producer_name, [])
+
+    def emit_df(
+        self,
+        pair_key: PairWithTimeframe,
+        dataframe: DataFrame
+    ) -> None:
+        """
+        Send this dataframe as an ANALYZED_DF message to RPC
+
+        :param pair_key: PairWithTimeframe tuple
+        :param data: Tuple containing the DataFrame and the datetime it was cached
+        """
+        if self.__rpc:
+            self.__rpc.send_msg(
+                {
+                    'type': RPCMessageType.ANALYZED_DF,
+                    'data': {
+                        'key': pair_key,
+                        'value': dataframe
+                    }
+                }
+            )
+
+    def add_external_df(
```

Comment:
Yes I was thinking that. I had trouble deciding to prefix it with a `_` because another object in the application calls it so it isn't necessarily private, but should be private to the user for sure. This method is automatically called by the ExternalMessageConsumer. So, the ApiServer makes available a websocket to listen to all RPCMessages from a bot. All the ExternalMessageConsumer does is listen for WHITELIST, and ANALYZED_DF messages from a Producer, and makes use of that data by adding it to the DataProvider for use in a Strategy.

Reply:
that's for sure a difficult call.
in principle, it's not private as others use it - but i do agree that for user-facing interfaces, we should adapt this a bit - making methods private where we expect the user to not use it.

otherwise it'd be public - but if a user really uses this, then it could be problematic as it's corrupting the external DF store.